### PR TITLE
C17 time frame

### PR DESCRIPTION
**Pull request by Jun0305**
Long click the "+" image in the meal plan and then the type fragment appears. Choose any recipe wanted to add to the meal plan

### CODE REVIEW

**Comment by nathanbillis** [comment]
[C17 - **Enable Meal Planner** As a user, I want to be able to select a time frame to see certain recipes, such as breakfast, lunch or dinner](https://trello.com/c/Ntsv5G1u/37-c17-enable-meal-planner-as-a-user-i-want-to-be-able-to-select-a-time-frame-to-see-certain-recipes-such-as-breakfast-lunch-or-din)

**Comment by profjb58** [comment]
Will have a look at Monday.

File: app/src/main/java/com/group4sweng/scranplan/Home.java [hunk 1]
```diff
@@ -513,9 +514,9 @@ public void initSearchMenu(){
         mScoreBox = layout.findViewById(R.id.scoreCheckBox);
         mVoteBox = layout.findViewById(R.id.voteCheckBox);
         mTimeBox = layout.findViewById(R.id.timestampCheckBox);
-        mIngredientsBox = layout.findViewById(R.id.ingredientCheckBox);
-        mNameBox = layout.findViewById(R.id.nameCheckBox);
-        mChefBox = layout.findViewById(R.id.chefCheckBox);
+        mIngredientsBox = layout.findViewById(R.id.BreackfastCheckBox);
```
Comment (nathanbillis) [inline app/src/main/java/com/group4sweng/scranplan/Home.java:517]: - Change to breakfastCheckBox

File: app/src/main/java/com/group4sweng/scranplan/Home.java [hunk 2]
```diff
@@ -513,9 +514,9 @@ public void initSearchMenu(){
...
+        mNameBox = layout.findViewById(R.id.LunchCheckBox);
```
Comment (nathanbillis) [inline app/src/main/java/com/group4sweng/scranplan/Home.java:518]: lunchCheckBox

File: app/src/main/java/com/group4sweng/scranplan/Home.java [hunk 3]
```diff
@@ -513,9 +514,9 @@ public void initSearchMenu(){
...
+        mChefBox = layout.findViewById(R.id.DinnerCheckBox);
```
Comment (nathanbillis) [inline app/src/main/java/com/group4sweng/scranplan/Home.java:519]: dinnerCheckBox

**Comment by profjb58** [comment]
Some serious issues with the following...

image
- null object reference when selecting a recipe for the mealplanner from a long press.
- By default adds in all available allergens and dietary requirements. Should not do this.
- Long press also serves the function of removing a recipe so when removing a recipe from the mealplanner this dialog pops up. This shouldn't happen.
- Overall not very intuitive, James will go over this i'm sure and we can add instructions on mealplanner initial launch.

In future look into writing espresso tests for this kind of thing. These should be added anyway and manual testing shouldn't really be an option but if your stuck I can help go through writing those.

File: app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java [hunk 4]
```diff
@@ -66,6 +66,54 @@ public void checkAddButton() {
...
+        onView(withId(2)).perform(longClick());
+
+        Thread.sleep(THREAD_SLEEP_TIME/4);
+
+        onView(withText("slow cooker spiced root and lentil casserole")).perform(click());
+        onView(withText("Add")).perform(click());
+
+        Thread.sleep(THREAD_SLEEP_TIME/4);
+    }
+
```
Comment (profjb58) [inline app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java:116]: Also need checks for the contents of what is retrieved when you hold press on a recipe. Hence why the issues I presented have surfaced.

File: app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java [hunk 5]
```diff
@@ -175,6 +198,96 @@ private void defaultButton(final ImageButton imageButton) {
...
+            mUser.setMealPlanner(plannerList);
+            updateMealPlan();
+        });
+
+        //Updates planner and user
+        plannerList.set(imageButton.getId(), null);
+        mUser.setMealPlanner(plannerList);
+        updateMealPlan();
+    }
+
```
Comment (profjb58) [inline app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java:290]: Lots of duplicate code. Can be simplified.

File: app/src/main/java/com/group4sweng/scranplan/SearchFunctions/SearchPrefs.java [hunk 6]
```diff
@@ -23,12 +23,16 @@
...
+    Boolean mBreakfast;
+    Boolean mLunch;
+    Boolean mDinner;
 
     // Building a new search preferences following user input
     public SearchPrefs(Boolean pes, Boolean vegi, Boolean vega, Boolean nut,
                        Boolean milk, Boolean egg, Boolean wheat, Boolean shell,
                        Boolean soy, Boolean score, Boolean vote, Boolean time,
-                       Boolean ingred, Boolean name, Boolean chef){
+                       Boolean ingred, Boolean name, Boolean chef/*, Boolean breakfast,
```
Comment (profjb58) [inline app/src/main/java/com/group4sweng/scranplan/SearchFunctions/SearchPrefs.java:34]: ingred > ingredient, weird naming scheme.

File: app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java (hunk 4 above)
Comment (profjb58) [reply app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java:116]: Also dosen't acctually check anything. Just performs an automated task. Look into using 'assert' conditions from JUnit or espresso condition checks.

**Comment by profjb58** [comment]
To be honest I don't really understand why there are 3 seperate adapters and views for the dinner, lunch and breakfast stuff as well. Having a single fragment or activity that acts for all 3 and just adjusts the name and recipes based on whats available makes much more sense. I would highly reccomend refactoring to fit this but it depends on what @nathanbillis thinks.

File: app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java [hunk 7]
```diff
@@ -46,6 +47,9 @@
     //Fragment handlers
     private FragmentTransaction fragmentTransaction;
     private RecipeFragment recipeFragment;
+    private BreakfastFragment breakfastFragment;
+    private LunchFragment lunchFragment;
+    private DinnerFragment dinnerFragment;
```
Comment (nathanbillis) [inline app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java:52]: private MealTimescaleFragment dinnerFragment;

**Comment by nathanbillis** [comment]
> 
> 
> To be honest I don't really understand why there are 3 seperate adapters and views for the dinner, lunch and breakfast stuff as well.
> The lines of code added are +2056 because there is a lot of duplicate code as a result of this.
> 
> Having a single fragment or activity that acts for all 3 and just adjusts the name and recipes based on whats available makes much more sense. I would highly reccomend refactoring to fit this or re-doing this story from scratch.
> 
> Apologies if what i've posted seems a bit harsh, trying to be to the point, there are some really good features and it's good that the majority of it works. I hope you can see where i'm coming from though.

I also agree with this, I'm a bit confused as to why there are 3 pieces that do the same thing. Couldn't we have this working with two classes one for query and the other for recycler view, so the breakfast fragment, lunch fragment, and dinner fragment could be chaged to MealTimescaleFragement because the code in all the classes is the same.

File: app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java [hunk 8]
```diff
@@ -66,6 +66,21 @@ public void checkAddButton() {
...
+    public void BreakfastAddToPlanner() throws InterruptedException {
+        Log.d(TAG, "Testing searching");
+
+        onView(withText("Meal Planner")).perform(click());
+        onView(withId(0)).perform(click());
+
+        Thread.sleep(THREAD_SLEEP_TIME/4);
+
+        onView(withText("Soup")).perform(click());
+        onView(withText("Add")).perform(click());
```
Comment (profjb58) [inline app/src/androidTest/java/com/group4sweng/scranplan/PlannerFragmentTest.java:81]: This test may fail if the 'soup' recipe is not visible.
Also dosen't guarentee that you are clicking on the 'lunch', 'dinner' or 'breakfast' segment.

File: app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java [hunk 9]
```diff
@@ -146,20 +157,34 @@ private void openRecipeDialog(SearchQuery query) {
...
+                lunchFragment.setTargetFragment(PlannerFragment.this, lunchRequestCode);
+                fragmentTransaction = getParentFragmentManager().beginTransaction();
+                fragmentTransaction.add(R.id.frameLayout, lunchFragment); //Overlays fragment on existing one
+            }else if (i == 2){
+                dinnerFragment = new DinnerFragment(mUser);
+                dinnerFragment.setArguments(bundle);
+                dinnerFragment.setTargetFragment(PlannerFragment.this, dinnerRequestCode);
+                fragmentTransaction = getParentFragmentManager().beginTransaction();
+                fragmentTransaction.add(R.id.frameLayout, dinnerFragment); //Overlays fragment on existing one
+            }
```
Comment (profjb58) [inline app/src/main/java/com/group4sweng/scranplan/MealPlanner/PlannerFragment.java:187]: Still really confused why we are using 3 fragments for basically the same thing with different text + images.
It does work though I guess.

**Comment by nathanbillis** [comment]
Tests are failing:

 [Image from Gyazo](https://gyazo.com/7a76cc9390bb31da42a9b4ded03cf059)

**Comment by Jun0305** [comment]
have merged three fragments into one

**Comment by profjb58** [comment]
Jun has moved stuff to the RecipeFragment which has greatly simplified things, just gonna add Espresso tests and check functionality before reccomending a pull request.
Very busy atm so I can't promise much but i'll try to do asap.

**Comment by nathanbillis** [comment]
@profjb58 @Jun0305 is this ready to go yet?

**Comment by profjb58** [comment]
Branch had over 15+ file conflicts, to try and understand how to rebase this and add it in would be near impossible.
You are welcome to try Jun by reverting the rebase I did and having a go yourself but i'm not sure we will ever be able to merge this.

It also has the potential to break work done by Nathan in fixing the searching in the mealplanner, and work on queries which has changed a bit since.

**Comment by profjb58** [comment]
Alright. So this entire branch can be implemented through the following steps Jun.

- Check if a recipe is for lunch breakfast or dinner. You don't need to use queries, the information is within the Firebase database under the 'recipes' collection as boolean values.
_(I imagine this might have already been used for this story but the following describes an easier solution)_
- Simply check when the 'Top Picks', 'Trending' and 'New tastes' scrollviews are populated that checks are made for the breakfast, lunch and dinner variable.

**Comment by nathanbillis** [comment]
PR closed, Superseded by #110

**Pull request closed without merging**